### PR TITLE
Port Count

### PR DESCRIPTION
**Pull request by iamemilio**
Allows you to specify how many ports to attach to each machine created for a given network.
This patch also includes some fixes to how ports are attached to nodes:
- ensures no duplicate ports
- ensures ports are not taken
- ensures ports meet networking needs specified by user

Fixes: [OSASINFRA-2364](https://issues.redhat.com/browse/OSASINFRA-2364)

### CODE REVIEW

**Comment by pierreprinetti** [comment]
I know it feels like overhead, but I would be grateful if, when possible, you could add a link to a Jira card (or a BZ) to PRs, so that it's easier for the bypasser to get some context

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go [hunk 1]
```diff
@@ -134,7 +134,17 @@ type NetworkParam struct {
 	NoAllowedAddressPairs bool `json:"noAllowedAddressPairs,omitempty"`
 	// PortTags allows users to specify a list of tags to add to ports created in a given network
 	PortTags []string `json:"portTags,omitempty"`
-	VNICType string   `json:"vnicType,omitempty"`
+	// VNICTYpe specifies the nic type of the ports created on this network
+	// Default: "Normal"
+	// +optional
+	VNICType string `json:"vnicType,omitempty"`
+	// PortCount is a positive, non-zero, intiger representing the number of ports in this network to attach to each node
```
Comment (Fedosin) [inline pkg/apis/openstackproviderconfig/v1alpha1/types.go:141]: integer

**Comment by iamemilio** [comment]
don't retest until Nova is back up on vexxhost

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go [hunk 2]
```diff
@@ -134,7 +134,17 @@ type NetworkParam struct {
...
+	// Default: 1
+	// +optional
+	PortCount uint `json:"portCount,omitempty"`
+	// PortSecurity is a boolean value that indicates whether security is enabled on ports created for this subnet
+	// Default: Nova default
+	PortSecurity *bool `json:"portSecurity,omitempty"`
```
Comment (Fedosin) [inline pkg/apis/openstackproviderconfig/v1alpha1/types.go:147]: For VNICType and PortCount you specified "+optional". Is PortSecurity a mandatory field?

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go [hunk 3]
```diff
@@ -134,7 +134,17 @@ type NetworkParam struct {
 	NoAllowedAddressPairs bool `json:"noAllowedAddressPairs,omitempty"`
 	// PortTags allows users to specify a list of tags to add to ports created in a given network
 	PortTags []string `json:"portTags,omitempty"`
-	VNICType string   `json:"vnicType,omitempty"`
+	// VNICTYpe specifies the nic type of the ports created on this network
```
Comment (Fedosin) [inline pkg/apis/openstackproviderconfig/v1alpha1/types.go:137]: VNICType

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go [hunk 4]
```diff
@@ -134,7 +134,17 @@ type NetworkParam struct {
...
+	// Default: "Normal"
```
Comment (Fedosin) [inline pkg/apis/openstackproviderconfig/v1alpha1/types.go:138]: You don't set the default in the code... Are you sure that it is always "Normal"?

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go (hunk 1 above)
Comment (iamemilio) [reply pkg/apis/openstackproviderconfig/v1alpha1/types.go:141]: lol oops, thanks

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go (hunk 2 above)
Comment (iamemilio) [reply pkg/apis/openstackproviderconfig/v1alpha1/types.go:147]: no, it will default to nova default. I will update that.

File: pkg/cloud/openstack/clients/machineservice.go [hunk 5]
```diff
@@ -421,35 +437,62 @@ func (is *InstanceService) InstanceCreate(clusterName string, name string, clust
 		if err != nil {
 			return nil, err
 		}
+
+		// Set default number of ports created per network to 1
+		portCount := 1
+		if net.PortCount > 1 {
```
Comment (dulek) [inline pkg/cloud/openstack/clients/machineservice.go:443]: 0 will default to 1?

File: Dockerfile [hunk 6]
```diff
@@ -4,6 +4,6 @@ COPY . .
 
 RUN go build -o ./machine-controller-manager ./cmd/manager
 
-FROM registry.ci.openshift.org/openshift/origin-v4.7:base
+FROM registry.ci.openshift.org/origin/4.8:base
```
Comment (EmilienM) [inline Dockerfile:7]: I'm curious why this change happens here (and why we don't use a separate commit for that?)

**Review by iamemilio — COMMENTED** [review]
f

File: Dockerfile (hunk 6 above)
Comment (iamemilio) [reply Dockerfile:7]: I should split it into its own commit

File: pkg/cloud/openstack/clients/machineservice.go [hunk 7]
```diff
@@ -421,35 +437,61 @@ func (is *InstanceService) InstanceCreate(clusterName string, name string, clust
...
-					vnicType:  net.VNICType,
-				})
+				// Create one NIC per count
+				for i := 0; i < portCount; i++ {
+					nets = append(nets, ServerNetwork{
+						networkID:    netID,
+						portTags:     net.PortTags,
+						vnicType:     net.VNICType,
+						portSecurity: net.PortSecurity,
+						portName:     fmt.Sprintf("%s-%d", name, i),
```
Comment (MaysaMacedo) [inline pkg/cloud/openstack/clients/machineservice.go:459]: I'm not sure where the instance name is enforced, but could it have the following format <infra-id>-master-<n>? and consequently port name <infra-id>-master-<n>-<n>?

Comment (iamemilio) [reply pkg/cloud/openstack/clients/machineservice.go:459]: hmm, I am not sure actually. Can CAPO manage control plane resources? If so, then this is a possibility.  For workers the name scheme is <cluster-id>-<machineset name>-<unique hash>

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go [hunk 8]
```diff
@@ -134,7 +134,18 @@ type NetworkParam struct {
 	NoAllowedAddressPairs bool `json:"noAllowedAddressPairs,omitempty"`
 	// PortTags allows users to specify a list of tags to add to ports created in a given network
 	PortTags []string `json:"portTags,omitempty"`
-	VNICType string   `json:"vnicType,omitempty"`
+	// VNICType specifies the nic type of the ports created on this network
```
Comment (adduarte) [inline pkg/apis/openstackproviderconfig/v1alpha1/types.go:137]: nic = vnic type. 
also comment seems redundant. the name vnicType seems clear enough :)

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go [hunk 9]
```diff
@@ -134,7 +134,18 @@ type NetworkParam struct {
...
+	// +optional
+	VNICType string `json:"vnicType,omitempty"`
+	// PortCount is a positive, non-zero, integer representing the number of ports in this network to attach to each node
+	// Default: 1
+	// +optional
+	PortCount uint `json:"portCount,omitempty"`
+	// PortSecurity is a boolean value that indicates whether security is enabled on ports created for this subnet
+	// Default: Nova default
+	// +optional
+	PortSecurity *bool `json:"portSecurity,omitempty"`
```
Comment (adduarte) [inline pkg/apis/openstackproviderconfig/v1alpha1/types.go:148]: I appriciate the comments, but unless they get consumed by something else (like IDE autocompletion, or soemthing like that), usually they end up being more work to mantain. Best if the code itself makes it obvious what it does and what variables are.

File: pkg/cloud/openstack/clients/machineservice.go [hunk 10]
```diff
@@ -95,11 +96,26 @@ type Instance struct {
...
-	vnicType  string
+	networkID    string
+	subnetID     string
+	portTags     []string
+	vnicType     string
+	portSecurity *bool
+	portName     string
 }
+
+// for geting vnic type when listing ports
```
Comment (adduarte) [inline pkg/cloud/openstack/clients/machineservice.go:107]: remove

File: pkg/cloud/openstack/clients/machineservice.go [hunk 11]
```diff
@@ -95,11 +96,26 @@ type Instance struct {
...
+type portWithPortsbinding struct {
+	ports.Port
+	portsbinding.PortsBindingExt
+}
+
+// for updating port security
```
Comment (adduarte) [inline pkg/cloud/openstack/clients/machineservice.go:113]: remove

File: pkg/cloud/openstack/clients/machineservice.go [hunk 12]
```diff
@@ -478,40 +520,83 @@ func (is *InstanceService) InstanceCreate(clusterName string, name string, clust
...
 
+	// Convert nets list into a list of servers.Network to be passed as NICs for instance create
 	userData := base64.StdEncoding.EncodeToString([]byte(cmd))
 	var ports_list []servers.Network
+	takenPort := map[string]bool{}
 	for _, net := range nets {
 		if net.networkID == "" {
 			return nil, fmt.Errorf("No network was found or provided. Please check your machine configuration and try again")
 		}
+
```
Comment (adduarte) [inline pkg/cloud/openstack/clients/machineservice.go:531]: gratuitous line

File: pkg/cloud/openstack/clients/machineservice.go [hunk 13]
```diff
@@ -478,40 +520,83 @@ func (is *InstanceService) InstanceCreate(clusterName string, name string, clust
...
 		if err != nil {
 			return nil, fmt.Errorf("Searching for existing port for server err: %v", err)
 		}
-		portList, err := ports.ExtractPorts(allPages)
+		var portList []portWithPortsbinding
+		err = ports.ExtractPortsInto(allPages, &portList)
 		if err != nil {
 			return nil, fmt.Errorf("Searching for existing port for server err: %v", err)
 		}
+
```
Comment (adduarte) [inline pkg/cloud/openstack/clients/machineservice.go:545]: gratuitous

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go (hunk 8 above)
Comment (iamemilio) [reply pkg/apis/openstackproviderconfig/v1alpha1/types.go:137]: The name needs to be in the comment as a go fmt requirement I believe. The Comments get consumed by automatic documentation generation tooling I believe

**Comment by mdbooth** [comment]
@iamemilio @mandre I just had a quick look in upstream CAPO and I didn't immediately see any SR-IOV support there.

I wonder if we should discuss this patch upstream first to ensure that:
* I didn't miss anything
* This approach will be acceptable there
* There's nobody else working on an incompatible solution which hasn't landed yet

It would be good to at least validate design changes like this, even if we don't do a full upstream port ourselves.

File: pkg/cloud/openstack/clients/machineservice.go (hunk 7 above)
Comment (iamemilio) [reply pkg/cloud/openstack/clients/machineservice.go:459]: Ok, it does, so this could be a problem. The way I see it, there are 2 ways to handle this:
1. Change the naming scheme of all instances to follow the `<cluster-id>-<machineset name>-<unique hash>` model. 
2. Only append a number at the end of a port's name if more than one port is requested, that way the only valid usage is for worker nodes (for now)

@maysamacedo I am not sure if either of these would have consequences for kuryr, but if either is acceptable, I would lean towards 1.

**Comment by iamemilio** [comment]
This approach will likely not be acceptable upstream unless they have support for OpenShift Operators. I would need to follow up with NFV to see if there is any upstream equivilent of the OpenShift SR-IOV Operator for upstream kubernetes.

File: pkg/apis/openstackproviderconfig/v1alpha1/types.go (hunk 9 above)
Comment (iamemilio) [reply pkg/apis/openstackproviderconfig/v1alpha1/types.go:148]: they do :0

File: pkg/cloud/openstack/clients/machineservice.go (hunk 7 above)
Comment (iamemilio) [reply pkg/cloud/openstack/clients/machineservice.go:459]: Ok, I thought about this, and while its ugly for now, I would rather just leave it and make it a consideration in a bz we are going to file to eventually have upstream parity with this feature based on this work happening upstream: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/778

We will fix this post FF by allowing the port names to be explicitly setable.

Comment (MaysaMacedo) [reply pkg/cloud/openstack/clients/machineservice.go:459]: We're working on moving away from relying on ports names on 4.8, so it's fine. I was just not sure is the name of that Port would be something like `ostest-qxzjd-master-port-0-3` and if it that would be expected.

Comment (iamemilio) [reply pkg/cloud/openstack/clients/machineservice.go:459]: yeah, that would be a possibility for the master nodes. Its ugly, but we are going to fix it by letting you specify port names when we adopt https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/778

**Comment by iamemilio** [comment]
Tested in my local env, and this works fine. There seems to be an issue with the CI

**Comment by iamemilio** [comment]
Closing in favor of https://github.com/openshift/cluster-api-provider-openstack/pull/173
